editoast/src/views/scenario/macro_nodes.rs

```diff
@@ -0,0 +1,614 @@
+use crate::models::Scenario;
+use crate::models::Study;
+use crate::views::pagination::PaginatedList;
+use crate::views::pagination::PaginationQueryParam;
+use crate::views::pagination::PaginationStats;
+use crate::views::projects::ProjectError;
+use crate::views::projects::ProjectIdParam;
+use crate::views::scenario::ScenarioError;
+use crate::views::scenario::ScenarioIdParam;
+use crate::views::study::StudyError;
+use crate::views::study::StudyIdParam;
+use crate::views::AuthorizationError;
+use crate::views::AuthorizerExt;
+
+crate::routes! {
+    "/macro_nodes" => {list, create,},
+    "/macro_nodes/{node_id}" => {get, update, delete,},
+}
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "macro_node")]
+pub enum MacroNodeError {
+    #[error("Node '{node_id}', could not be found")]
+    #[editoast_error(status = 404)]
+    NotFound { node_id: i64 },
+}
+
+#[derive(IntoParams, Deserialize)]
+#[allow(unused)]
+pub struct MacroNodeIdParam {
+    node_id: i64,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, ToSchema)]
+pub struct MacroNodeForm {
+    pub position_x: f64,
+    pub position_y: f64,
+    pub full_name: Option<String>,
+    pub connection_time: i64,
+    pub labels: Vec<Option<String>>,
+    pub trigram: Option<String>,
+    pub path_item_key: String,
+}
+
+impl MacroNodeForm {
+    pub fn into_macro_node_changeset(self, scenario_id: i64) -> Result<Changeset<MacroNode>> {
+        let macro_node = MacroNode::changeset()
+            .scenario_id(scenario_id)
+            .position_x(self.position_x)
+            .position_y(self.position_y)
+            .full_name(self.full_name)
+            .connection_time(self.connection_time)
+            .labels(self.labels)
+            .trigram(self.trigram)
+            .path_item_key(self.path_item_key);
+
+        Ok(macro_node)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, ToSchema)]
+pub struct MacroNodeResponse {
+    pub id: i64,
+    pub position_x: f64,
+    pub position_y: f64,
+    pub full_name: Option<String>,
+    pub connection_time: i64,
+    pub labels: Vec<Option<String>>,
+    pub trigram: Option<String>,
+    pub path_item_key: String,
+}
+
+impl MacroNodeResponse {
+    pub fn new(node: MacroNode) -> Self {
+        Self {
+            id: node.id,
+            position_x: node.position_x,
+            position_y: node.position_y,
+            full_name: node.full_name,
+            connection_time: node.connection_time,
+            labels: node.labels,
+            trigram: node.trigram,
+            path_item_key: node.path_item_key,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, PartialEq, Clone, ToSchema)]
+#[cfg_attr(test, derive(Deserialize))]
+pub struct MacroNodeListResponse {
+    #[serde(flatten)]
+    stats: PaginationStats,
+    results: Vec<MacroNodeResponse>,
+}
+
+/// Get macro node list by scneario id
+#[utoipa::path(
+    get, path = "",
+    tag = "scenarios, nge",
```

Comment:
Would prefer to not use the term "NGE" in our APIs. "NGE" is the tool, "macro_nodes" is the object type.

Reply:
All the endpoints already have `macro_node` in the url, there is no need to add it as a tag on the openapi definition.  
So I remove the `nge` tag on each openapi definition (see your next comments on `generatedEditoastApi.ts` and `ngraph`)